autoPyTorch/api/results_manager.py

```diff
@@ -55,42 +82,127 @@ def _extract_metrics_info(
 
     if run_value.status not in (StatusType.SUCCESS, StatusType.DONOTADVANCE):
         # Additional info for metrics is not available in this case.
-        return {metric.name: np.nan for metric in scoring_functions}
+        return {metric.name: metric._worst_possible_result for metric in scoring_functions}
 
     cost_info = run_value.additional_info['opt_loss']
     avail_metrics = cost_info.keys()
 
     return {
         metric.name: cost2metric(cost=cost_info[metric.name], metric=metric)
-        if metric.name in avail_metrics else np.nan
+        if metric.name in avail_metrics else metric._worst_possible_result
         for metric in scoring_functions
     }
 
 
+class EnsembleResults:
+    def __init__(
+        self,
+        metric: autoPyTorchMetric,
+        ensemble_performance_history: List[Dict[str, Any]],
+        order_by_endtime: bool = False
+    ):
+        self._test_scores: List[float] = []
+        self._train_scores: List[float] = []
+        self._end_times: List[float] = []
+        self._metric = metric
+
+        self._extract_results_from_ensemble_performance_history(ensemble_performance_history)
+        if order_by_endtime:
+            self.sort_by_endtime()
+
+    @property
+    def train_scores(self) -> np.ndarray:
+        return np.asarray(self._train_scores)
+
+    @property
+    def test_scores(self) -> np.ndarray:
+        return np.asarray(self._test_scores)
+
+    @property
+    def end_times(self) -> np.ndarray:
+        return np.asarray(self._end_times)
+
+    def update(self, train_score: float, test_score: float, end_time: float) -> None:
+        self._train_scores.append(train_score)
+        self._test_scores.append(test_score)
+        self._end_times.append(end_time)
+
+    def clear(self) -> None:
+        self._test_scores = []
+        self._train_scores = []
+        self._end_times = []
+
+    def sort_by_endtime(self) -> None:
+        """
+        Since the default order is by start time
+        and parallel computation might change the order of ending,
+        this method provides the feature to sort by end time.
+        Note that this method is destructive.
+        """
+        order = np.argsort(self._end_times)
+
+        self._train_scores = self.train_scores[order].tolist()
+        self._test_scores = self.test_scores[order].tolist()
+        self._end_times = self.end_times[order].tolist()
+
+    def _extract_results_from_ensemble_performance_history(
+        self,
+        ensemble_performance_history: List[Dict[str, Any]]
+    ) -> None:
+        """
+        Extract the information to match this class format.
+
+        Args:
+            ensemble_performance_history (List[Dict[str, Any]]):
+                The history of the ensemble optimization from SMAC.
```

Comment:
Ensemble optimization is not happening in SMAC. This information comes from [here](https://github.com/automl/Auto-PyTorch/blob/ef8d21ae982883548a8e77a3690eefa810d3ae07/autoPyTorch/ensemble/ensemble_builder.py#L856) in `EnsembleBuilder`.

Reply:
Oh, thanks for the catch.